File: packages/backend/src/http/providers.ts
import { publicProvider } from '../vault/public-provider'
import {
  AccountRefused,
  type ProviderAccounts
} from '../vault/provider-accounts'
import type { ProviderOperations } from '../vault/provider-operations'
import { errorMessage } from '@demicodes/utils'
import { Hono, type Context } from 'hono'
import { z } from 'zod'
import { publicQuota } from '../llm/provider-details'
import { configuredModelsSchema } from '../llm/model-config'
import type { AuthEnv, InstanceMode } from '../auth/identity'
import type { ProviderAssembly, SessionProviderContext } from '../llm/assembly'
import type { CliInstalls } from '../llm/cli-installs'
import { booleanQuerySchema } from './query'
import type { VendorCatalog } from '../llm/vendors'
import { canConfigureProviders } from '../vault/scope'
import type {
  ApiKeyProviderConfig,
  ProviderEntry,
  ProviderVault
} from '../vault/providers'
import type { SubscriptionLoginFlows } from '../vault/subscription-login'

const subscriptionLoginBodySchema = z.strictObject({
  providerType: z.string().min(1),
  label: z.string().min(1).optional(),
})


/**
 * An API-key entry comes in two shapes: from the vendor catalog — the
 * vendor's family and endpoint prefilled, its model list live unless the
 * user types one — or a custom endpoint naming the family itself.
 */
const createProviderBodySchema = z.union([
  z.strictObject({
    vendorId: z.string().min(1),
    label: z.string().min(1),
    apiKey: z.string().min(1),
    baseUrl: z.url().optional(),
    models: configuredModelsSchema.optional(),
  }),
  z.strictObject({
    providerType: z.string().min(1),
    label: z.string().min(1),
    apiKey: z.string().min(1),
    wireApi: z.enum(['responses', 'chat-completions']).optional(),
    baseUrl: z.url().optional(),
    models: configuredModelsSchema.optional(),
  }),
])

/**
 * Edits: the label of any entry; endpoint, key and model list of an API-key
 * entry (`models: null` returns to the live list).
 */
const patchProviderBodySchema = z.strictObject({
  label: z.string().min(1).optional(),
  apiKey: z.string().min(1).optional(),
  baseUrl: z.url().nullable().optional(),
  models: configuredModelsSchema.nullable().optional(),
})

/**
 * `/api/providers` — the vault surface in the caller's provider scope
 * (`vault/scope.ts`): the instance's providers in shared mode, where only
 * admins configure providers; the caller's own in isolated mode. Responses
 * never carry key material: an entry lists as its family, label, endpoint,
 * vendor and model list.
 */
export function providerRoutes(options: {
  accounts: ProviderAccounts
  operations: ProviderOperations
  vault: ProviderVault
  assembly: ProviderAssembly
  vendors: VendorCatalog
  logins: SubscriptionLoginFlows
  mode: InstanceMode
  /** A process provider's CLI: its versions and where it is installed (`claude-cli.md`). */
  cli: {
    installs: CliInstalls
    /** The vendor's newest version; `refresh` asks the vendor at once. */
    newest(refresh: boolean): Promise<string>
    /** Refuses a version the vendor does not publish. */
    verify(version: string): Promise<void>
    /** What each machine that can be asked now has installed. */
    machines(userId: string, signal: AbortSignal): Promise<Array<{
      deviceId: string
      name: string
      versions: string[] | null
    }>>
  }
  /** The machine for a process provider's work that belongs to no conversation. */
  placeAccountProcess: (
    userId: string,
    entry: ProviderEntry,
    signal: AbortSignal
  ) => Promise<{
    session: SessionProviderContext
    cwd: string
    release(): void
  }>
}): Hono<AuthEnv> {
  const { vault, assembly, vendors, logins, mode } = options
  const app = new Hono<AuthEnv>()
  app.onError((error, c) => {
    if (error instanceof AccountRefused)
      return c.json({ code: error.code, message: error.message }, error.status)
    throw error
  })

  const ownerOf = (c: Context<AuthEnv>) => vault.ownerFor(c.get('user').id)
  // A provider outside the caller's scope answers like a missing one.
  const scoped = async (c: Context<AuthEnv>) => {
    const provider = await vault.get(c.req.param('id') ?? '')
    return provider && provider.ownerUserId === await ownerOf(c) ? provider : null
  }
  // An account of a process provider is no use without its CLI, so adding
  // one starts the install where account work runs; nobody waits for it.
  const installCli = async (c: Context<AuthEnv>, entry: ProviderEntry) => {
    const provider = (await assembly.providerFor(entry.id))?.provider
    if (provider?.requiresProcessCapableHost)
      options.cli.installs.start(c.get('user').id, entry)
  }
  const accountBodySchema = z.object({
    credentialId: z.string().min(1).optional()
  })
  const invalidBody = (c: Context<AuthEnv>, error: z.ZodError) => {
    const issue = error.issues[0]
    return c.json(
      {
        code: 'invalid_body',
        message: `Invalid request body${issue ? `: ${issue.path.join('.')} ${issue.message}` : ''}`
      },
      400,
    )
  }

  // Configuring providers — creating, editing, logging in, testing, deleting —
  // is the master's in shared mode and everyone's own in isolated mode.
  app.use('*', async (c, next) => {
    const refreshingQuota = c.req.method === 'POST' &&
      /^\/api\/providers\/[^/]+\/quota$/.test(c.req.path)
    if (c.req.method !== 'GET' &&
      !canConfigureProviders(mode, c.get('user').role)) {
      return c.json({
        code: 'forbidden',
        message: 'Providers are configured by the instance owner'
      }, 403)
    }
    const id = c.req.path.split('/')[3]
    const providerMutation = id &&
      id !== 'subscription-login' &&
      id !== 'setup-token' &&
      c.req.method !== 'GET' &&
      !refreshingQuota &&
      !c.req.path.endsWith('/accounts/login')
    const release = providerMutation ? options.operations.reserve(id) : () => {}
    if (!release)
      return c.json({
        code: 'provider_busy',
        message: 'Another provider operation is still running'
      }, 409)
    try {
      await next()
    } finally {
      release()
    }
  })

  // Literal paths are registered before `/:id` so they win.

  /**
   * What the page can add: the models.dev vendors our runtimes speak to, and
   * each subscription family with whether the scope holds it.
   */
  app.get('/catalog', async (c) => {
    const entries = await vault.list({ ownerUserId: await ownerOf(c) })
    const subscriptions = assembly.typesOf('subscription').map(
      (providerType) => ({
        providerType,
        configured: entries.some(
          (entry) => entry.config.providerType === providerType
        ),
      })
    )
    try {
      return c.json({ subscriptions, vendors: await vendors.list() })
    } catch (error) {
      return c.json({
        code: 'catalog_unavailable',
        message: errorMessage(error)
      }, 502)
    }
  })

  app.post('/subscription-login', async (c) => {
    const parsed = subscriptionLoginBodySchema.safeParse(
      await c.req.json().catch(() => null)
    )
    if (!parsed.success)
      return c.json({
        code: 'invalid_body',
        message: 'Expected { providerType, label? }'
      }, 400)
    const { providerType } = parsed.data
    if (!assembly.credentialOf(providerType)) {
      return c.json({
        code: 'unknown_provider_type',
        message: `Unknown provider type "${providerType}"`
      }, 400)
    }
    const started = await logins.start(
      providerType,
      parsed.data.label ?? `${providerType} subscription`,
      await ownerOf(c)
    )
    if ('refused' in started) {
      if (started.refused === 'busy')
        return c.json({
          code: 'provider_busy',
          message: 'Provider login is already running'
        }, 409)
      return started.refused === 'exists'
        ? c.json({
          code: 'provider_exists',
          message: `This scope already has a ${providerType} subscription`
        }, 409)
        : c.json({
          code: 'no_login_flow',
          message: `Provider type "${providerType}" has no native login flow`
        }, 400)
    }
    return c.json({ login: { id: started.id, status: 'pending' } }, 202)
  })

  app.get('/subscription-login/:id', async (c) => {
    const state = logins.status(c.req.param('id') ?? '', await ownerOf(c))
    if (!state)
      return c.json({
        code: 'login_not_found',
        message: 'No such login flow'
      }, 404)
    return c.json({ login: state })
  })

  app.delete('/subscription-login/:id', async (c) => {
    if (!(await logins.cancel(c.req.param('id'), await ownerOf(c))))
      return c.json({
        code: 'login_not_found',
        message: 'No such login flow'
      }, 404)
    return c.body(null, 204)
  })

  app.post('/setup-token', async (c) => {
    const parsed = z.strictObject({
      token: z.string().trim().min(1).max(16384),
      label: z.string().trim().min(1).max(80)
    }).safeParse(await c.req.json().catch(() => null))
    if (!parsed.success)
      return c.json({
        code: 'invalid_body',
        message: 'Expected { token, label }'
      }, 400)
    const entry = await options.accounts.importClaude(
      await ownerOf(c),
      parsed.data.label,
      parsed.data.token
    )
    await installCli(c, entry)
    return c.json({ provider: publicProvider(entry) }, 201)
  })

  app.post('/:id/accounts/login', async (c) => {
    const entry = await scoped(c)
    if (!entry)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    if (entry.config.kind !== 'subscription')
      return c.json({
        code: 'unsupported',
        message: 'This provider does not use subscription accounts'
      }, 400)
    const started = await logins.start(
      entry.config.providerType,
      entry.label,
      await ownerOf(c),
      entry
    )
    if ('refused' in started)
      return c.json({
        code: started.refused,
        message: 'Device login is unavailable or already running; Claude uses setup-token import'
      }, 409)
    return c.json({ login: { id: started.id, status: 'pending' } }, 202)
  })

  app.get('/:id/accounts', async (c) => {
    const entry = await scoped(c)
    if (!entry)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    return c.json(await options.accounts.list(entry))
  })

  app.post('/:id/accounts', async (c) => {
    const entry = await scoped(c)
    if (!entry)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    const parsed = z.strictObject({ token: z.string().trim().min(1).max(16384) })
      .safeParse(await c.req.json().catch(() => null))
    if (!parsed.success)
      return c.json({
        code: 'invalid_body',
        message: 'Expected { token }'
      }, 400)
    const added = await options.accounts.addToken(entry, parsed.data.token)
    await installCli(c, entry)
    return c.json(added, 201)
  })

  app.put('/:id/accounts/active', async (c) => {
    const entry = await scoped(c)
    if (!entry)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    const parsed = z.strictObject({ credentialId: z.string().min(1) })
      .safeParse(await c.req.json().catch(() => null))
    if (!parsed.success)
      return c.json({
        code: 'invalid_body',
        message: 'Expected { credentialId }'
      }, 400)
    return c.json(
      await options.accounts.activate(entry, parsed.data.credentialId)
    )
  })

  app.delete('/:id/accounts/:credentialId', async (c) => {
    const entry = await scoped(c)
    if (!entry)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    await options.accounts.remove(entry, c.req.param('credentialId'))
    return c.body(null, 204)
  })

  app.get('/', async (c) => {
    const providers = await vault.list({ ownerUserId: await ownerOf(c) })
    return c.json({ providers: providers.map(publicProvider) })
  })

  app.post('/', async (c) => {
    const parsed = createProviderBodySchema.safeParse(
      await c.req.json().catch(() => null)
    )
    if (!parsed.success)
      return invalidBody(c, parsed.error)
    const body = parsed.data
    let config: ApiKeyProviderConfig
    if ('vendorId' in body) {
      const vendor = await vendors.get(body.vendorId)
      if (!vendor)
        return c.json({
          code: 'unknown_vendor',
          message: `Unknown vendor "${body.vendorId}"`
        }, 400)
      const baseUrl = body.baseUrl ?? vendor.baseUrl ?? undefined
      config = {
        kind: 'api_key',
        providerType: vendor.providerType,
        apiKey: body.apiKey,
        vendorId: vendor.id,
        ...(vendor.wireApi ? { wireApi: vendor.wireApi } : {}),
        ...(baseUrl ? { baseUrl } : {}),
        ...(body.models ? { models: body.models } : {}),
      }
    } else {
      const credential = assembly.credentialOf(body.providerType)
      if (!credential) {
        return c.json({
          code: 'unknown_provider_type',
          message: `Unknown provider type "${body.providerType}"`
        }, 400)
      }
      if (credential !== 'api_key') {
        return c.json({
          code: 'subscription_only',
          message: `Provider type "${body.providerType}" is configured by its login flow`
        }, 400)
      }
      config = {
        kind: 'api_key',
        providerType: body.providerType,
        apiKey: body.apiKey,
        ...(body.wireApi ? { wireApi: body.wireApi } : {}),
        ...(body.baseUrl ? { baseUrl: body.baseUrl } : {}),
        ...(body.models ? { models: body.models } : {}),
      }
    }
    const provider = await vault.create({
      ownerUserId: await ownerOf(c),
      label: body.label,
      config
    })
    return c.json({ provider: publicProvider(provider) }, 201)
  })

  app.patch('/:id', async (c) => {
    const provider = await scoped(c)
    if (!provider)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    const parsed = patchProviderBodySchema.safeParse(
      await c.req.json().catch(() => null)
    )
    if (!parsed.success)
      return invalidBody(c, parsed.error)
    const body = parsed.data
    const editsConfig = body.apiKey !== undefined ||
      body.baseUrl !== undefined ||
      body.models !== undefined
    if (editsConfig && provider.config.kind !== 'api_key') {
      return c.json({
        code: 'subscription_only',
        message: 'A subscription entry only takes a new label'
      }, 400)
    }
    const config: ApiKeyProviderConfig | undefined =
      editsConfig && provider.config.kind === 'api_key'
        ? {
          ...provider.config,
          ...(body.apiKey !== undefined ? { apiKey: body.apiKey } : {}),
          ...(body.baseUrl !== undefined
            ? { baseUrl: body.baseUrl ??
                undefined }
            : {}),
          ...(body.models !== undefined
            ? { models: body.models ??
                undefined }
            : {}),
        }
        : undefined
    const updated = await vault.update(provider.id, {
      ...(body.label !== undefined ? { label: body.label } : {}),
      ...(config ? { config } : {}),
    })
    if (!updated)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    await assembly.invalidate(provider.id)
    return c.json({ provider: publicProvider(updated) })
  })

  app.delete('/:id', async (c) => {
    const provider = await scoped(c)
    if (!provider)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    await vault.delete(provider.id)
    await assembly.deleteProviderState(provider.id)
    options.cli.installs.forget(provider.id)
    return c.body(null, 204)
  })

  app.get('/:id/status', async (c) => {
    const entry = await scoped(c)
    if (!entry)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    const resolved = await assembly.providerFor(entry.id)
    if (!resolved)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    try {
      return c.json({
        providerId: entry.id,
        ...await assembly.details(entry, resolved.provider)
      })
    } catch (error) {
      return c.json({
        code: 'provider_status_failed',
        message: errorMessage(error)
      }, 502)
    }
  })

  app.post('/:id/quota', async (c) => {
    const entry = await scoped(c)
    if (!entry)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    const parsed = accountBodySchema.safeParse(
      await c.req.json().catch(() => ({}))
    )
    if (!parsed.success)
      return invalidBody(c, parsed.error)
    // The account the user names, or the one the entry infers with.
    const provider = parsed.data.credentialId
      ? await assembly.forAccount(entry, parsed.data.credentialId)
      : (await assembly.providerFor(entry.id))?.provider
    if (!provider)
      return c.json({
        code: 'account_not_found',
        message: 'No such account'
      }, 404)
    const quota = provider.quota
    const capability = quota?.capability()
    if (!quota || capability?.mode !== 'supported' || !capability.canProbe)
      return c.json({ quota: publicQuota(quota?.latest() ?? null) })
    if (capability.probeCost !== 'free')
      return c.json({
        code: 'quota_requires_inference',
        message: 'This provider cannot refresh quota without an inference request'
      }, 409)
    try {
      return c.json(
        { quota: publicQuota(
            await quota.probe({ force: true, signal: c.req.raw.signal })
          ) }
      )
    } catch (error) {
      return c.json({
        code: 'quota_unavailable',
        message: errorMessage(error)
      }, 502)
    }
  })

  app.get('/:id/cli', async (c) => {
    const entry = await scoped(c)
    if (!entry || entry.config.kind !== 'subscription')
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    const refresh = booleanQuerySchema.safeParse(c.req.query('refresh'))
    if (!refresh.success)
      return c.json({
        code: 'invalid_query',
        message: 'refresh must be true or false'
      }, 400)
    const [newest, machines] = await Promise.all([
      options.cli.newest(refresh.data ?? false).then(
        version => ({ version }),
        error => ({ error: errorMessage(error) })
      ),
      options.cli.machines(c.get('user').id, c.req.raw.signal),
    ])
    return c.json({
      newest,
      held: entry.config.cliVersion ?? null,
      install: options.cli.installs.state(c.get('user').id, entry.id),
      machines,
    })
  })

  app.put('/:id/cli', async (c) => {
    const entry = await scoped(c)
    if (!entry || entry.config.kind !== 'subscription')
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    const parsed = z.strictObject({ held: z.string().min(1).max(64).nullable() })
      .safeParse(await c.req.json().catch(() => null))
    if (!parsed.success)
      return invalidBody(c, parsed.error)
    if (parsed.data.held) {
      try {
        await options.cli.verify(parsed.data.held)
      } catch (error) {
        return c.json({ code: 'unknown_version', message: errorMessage(error) }, 400)
      }
    }
    const { cliVersion: _held, ...config } = entry.config
    await vault.update(entry.id, {
      config: parsed.data.held ? { ...config, cliVersion: parsed.data.held } : config
    })
    await assembly.invalidate(entry.id)
    return c.json({ held: parsed.data.held })
  })

  app.post('/:id/cli/install', async (c) => {
    const entry = await scoped(c)
    if (!entry || entry.config.kind !== 'subscription')
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    options.cli.installs.start(c.get('user').id, entry)
    return c.json({ install: options.cli.installs.state(c.get('user').id, entry.id) }, 202)
  })

  app.post('/:id/test', async (c) => {
    const provider = await scoped(c)
    if (!provider)
      return c.json({
        code: 'provider_not_found',
        message: 'No such provider'
      }, 404)
    const parsed = z.object({
      modelId: z.string().min(1),
      credentialId: z.string().min(1).optional()
    }).safeParse(
      await c.req.json().catch(() => null)
    )
    if (!parsed.success)
      return c.json({
        code: 'invalid_body',
        message: 'Expected { modelId: string, credentialId?: string }: the model to test with'
      }, 400)
    // The test ran either way: a provider that refused is the answer, with
    // its reason, not a failure of this request.
    try {
      return c.json(await assembly.testProvider(provider.id, parsed.data.modelId, {
        ...(parsed.data.credentialId ? { credentialId: parsed.data.credentialId } : {}),
        placeProcess: entry => options.placeAccountProcess(
          c.get('user').id,
          entry,
          c.req.raw.signal
        ),
      }))
    } catch (error) {
      return c.json({ ok: false, message: errorMessage(error) })
    }
  })

  return app
}

File: crates/backend/src/edge/providers.rs
//! `/api/providers` (`web-api.md` § Model configuration and provider
//! inspection): the entries of the caller's scope, the vendors an entry can
//! be added from, and each entry's status, quota and test. Every user of a
//! scope reads it; configuring, testing and refreshing usage are the
//! master's on a shared instance and each owner's on an isolated one. No
//! answer carries key material.

use std::sync::Arc;

use axum::Json;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use demi_core::WireApi;
use demi_provider::quota::QuotaError;
use demi_provider::{Provider, Secret};
use demi_web_api::auth::UserDto;
use demi_web_api::error::ErrorCode;
use demi_web_api::ids::{CredentialId, ProviderId};
use demi_web_api::providers::{
    CreateProvider, CredentialKind, ProviderAnswer, ProviderDetails, ProviderPatch, Providers, QuotaAnswer,
    QuotaRequest, SubscriptionFamily, TestRequest, TestResult, VendorCatalog,
};
use demi_web_api::text::EndpointUrl;

use super::body::{JsonBody, OptionalJsonBody};
use super::error::ApiError;
use super::gate::AuthUser;
use crate::backend::Services;
use crate::llm::families::ProviderFamily;
use crate::shard::Shards;
use crate::vault::entries::{ApiKeyConfig, EntryCredential, ProviderEntry};
use crate::vault::operations::OperationGuard;

/// Refuses a user who does not configure the scope's entries.
pub(super) fn configures(services: &Services, user: &UserDto) -> Result<(), ApiError> {
    if services.vault.configures(user) {
        Ok(())
    } else {
        Err(ApiError::forbidden("Providers are configured by the instance owner"))
    }
}

/// The entry the path names, when it is one of the caller's scope; any
/// other answers like a missing one.
pub(super) async fn scoped(services: &Services, user: &UserDto, id: &str) -> Result<ProviderEntry, ApiError> {
    let id = ProviderId::try_from(id).map_err(|_| ApiError::provider_not_found())?;
    services
        .vault
        .visible(&user.id, &id)
        .await?
        .ok_or_else(ApiError::provider_not_found)
}

/// Holds the entry for one change.
pub(super) fn reserve(services: &Services, entry: &ProviderEntry) -> Result<OperationGuard, ApiError> {
    services
        .operations
        .reserve(&entry.id)
        .ok_or_else(ApiError::provider_busy)
}

pub(super) async fn list(
    State(services): State<Arc<Services>>,
    AuthUser(user): AuthUser,
) -> Result<Json<Providers>, ApiError> {
    let owner = services.vault.owner_for(&user.id).await?;
    let entries = services.vault.entries(owner).await?;
    Ok(Json(Providers {
        providers: entries.iter().map(ProviderEntry::dto).collect(),
    }))
}

/// What the page can add: each subscription family with whether the scope
/// holds its entry, and the models.dev vendors a family speaks to.
pub(super) async fn catalog(
    State(services): State<Arc<Services>>,
    AuthUser(user): AuthUser,
) -> Result<Json<VendorCatalog>, ApiError> {
    let owner = services.vault.owner_for(&user.id).await?;
    let entries = services.vault.entries(owner).await?;
    let assembly = &services.assembly;
    let subscriptions = assembly
        .families()
        .subscriptions()
        .map(|family| SubscriptionFamily {
            provider_type: family.to_owned(),
            configured: entries.iter().any(|entry| entry.family == family),
        })
        .collect();
    let vendors = assembly.vendors().vendors().await.map_err(catalog_unavailable)?;
    Ok(Json(VendorCatalog { subscriptions, vendors }))
}

fn catalog_unavailable(error: impl ToString) -> ApiError {
    ApiError::new(
        StatusCode::BAD_GATEWAY,
        ErrorCode::CatalogUnavailable,
        error.to_string(),
    )
}

fn unknown_family(family: &str) -> ApiError {
    ApiError::new(
        StatusCode::BAD_REQUEST,
        ErrorCode::UnknownProviderType,
        format!("Unknown provider type \"{family}\""),
    )
}

/// The key a body carries, as a credential: one line of text.
fn api_key(text: String) -> Result<Secret, ApiError> {
    Secret::try_from(text).map_err(|error| ApiError::invalid_body(format!("apiKey: {error}")))
}

/// The API-key family `name`, which must speak `wire_api` when the entry
/// names one.
fn api_key_family<'a>(
    services: &'a Services,
    name: &str,
    wire_api: Option<WireApi>,
) -> Result<&'a Arc<dyn ProviderFamily>, ApiError> {
    let family = services
        .assembly
        .families()
        .get(name)
        .ok_or_else(|| unknown_family(name))?;
    if family.credential() == CredentialKind::Subscription {
        return Err(ApiError::new(
            StatusCode::BAD_REQUEST,
            ErrorCode::SubscriptionOnly,
            format!("Provider type \"{name}\" is configured by its login"),
        ));
    }
    if let Some(wire) = wire_api.filter(|wire| !family.wires().contains(wire)) {
        return Err(ApiError::invalid_body(format!(
            "wireApi: the {name} family does not speak {wire}"
        )));
    }
    Ok(family)
}

/// Creates an API-key entry, from the vendor list or for a custom endpoint.
pub(super) async fn create(
    State(services): State<Arc<Services>>,
    AuthUser(user): AuthUser,
    JsonBody(body): JsonBody<CreateProvider>,
) -> Result<(StatusCode, Json<ProviderAnswer>), ApiError> {
    configures(&services, &user)?;
    let (family, label, config) = match body {
        CreateProvider::Vendor {
            vendor_id,
            label,
            api_key: key,
            base_url,
            models,
        } => {
            let vendor = services
                .assembly
                .vendors()
                .vendor(&vendor_id)
                .await
                .map_err(catalog_unavailable)?
                .ok_or_else(|| {
                    ApiError::new(
                        StatusCode::BAD_REQUEST,
                        ErrorCode::UnknownVendor,
                        format!("Unknown vendor \"{vendor_id}\""),
                    )
                })?;
            api_key_family(&services, &vendor.provider_type, vendor.wire_api)?;
            let base_url =
                match base_url {
                    Some(endpoint) => Some(endpoint),
                    None => vendor.base_url.map(EndpointUrl::try_from).transpose().map_err(|_| {
                        catalog_unavailable(format!("models.dev names no usable endpoint for {vendor_id}"))
                    })?,
                };
            let config = ApiKeyConfig {
                api_key: api_key(key)?,
                base_url,
                wire_api: vendor.wire_api,
                vendor_id: Some(vendor.id),
                models,
            };
            (vendor.provider_type, label, config)
        }
        CreateProvider::Custom {
            provider_type,
            wire_api,
            label,
            api_key: key,
            base_url,
            models,
        } => {
            api_key_family(&services, &provider_type, wire_api)?;
            let config = ApiKeyConfig {
                api_key: api_key(key)?,
                base_url,
                wire_api,
                vendor_id: None,
                models,
            };
            (provider_type, label, config)
        }
    };
    let owner = services.vault.owner_for(&user.id).await?;
    let entry = services
        .vault
        .create_api_key(owner, family, label.into_string(), config)
        .await?;
    Ok((StatusCode::CREATED, Json(ProviderAnswer { provider: entry.dto() })))
}

/// Edits an entry: the label of any, and the key, endpoint and model list of
/// an API-key entry. A configuration edit starts the entry's provider and
/// catalog afresh.
pub(super) async fn update(
    State(services): State<Arc<Services>>,
    AuthUser(user): AuthUser,
    Path(id): Path<String>,
    JsonBody(patch): JsonBody<ProviderPatch>,
) -> Result<Json<ProviderAnswer>, ApiError> {
    configures(&services, &user)?;
    let entry = scoped(&services, &user, &id).await?;
    let _held = reserve(&services, &entry)?;
    let reconfigures = patch.api_key.is_some() || patch.base_url.is_some() || patch.models.is_some();
    let config = if reconfigures {
        let EntryCredential::ApiKey(config) = &entry.credential else {
            return Err(ApiError::new(
                StatusCode::BAD_REQUEST,
                ErrorCode::SubscriptionOnly,
                "A subscription entry takes only a new label",
            ));
        };
        let mut config = config.clone();
        if let Some(key) = patch.api_key {
            config.api_key = api_key(key)?;
        }
        if let Some(base_url) = patch.base_url {
            config.base_url = base_url;
        }
        if let Some(models) = patch.models {
            config.models = models;
        }
        Some(config)
    } else {
        None
    };
    let label = patch.label.map(|label| label.into_string());
    let updated = services
        .vault
        .update(entry.id.clone(), label, config)
        .await?
        .ok_or_else(ApiError::provider_not_found)?;
    if reconfigures {
        services.assembly.invalidate(&entry.id).await?;
    }
    Ok(Json(ProviderAnswer {
        provider: updated.dto(),
    }))
}

/// Deletes an entry with its accounts, catalog and quota snapshots.
pub(super) async fn delete(
    State(services): State<Arc<Services>>,
    AuthUser(user): AuthUser,
    Path(id): Path<String>,
) -> Result<StatusCode, ApiError> {
    configures(&services, &user)?;
    let entry = scoped(&services, &user, &id).await?;
    let _held = reserve(&services, &entry)?;
    services.vault.delete(entry.id.clone()).await?;
    services.assembly.forget(&entry.id).await?;
    Ok(StatusCode::NO_CONTENT)
}

/// The entry's health, accounts and quota; reading it never probes and
/// never runs inference.
pub(super) async fn status(
    State(services): State<Arc<Services>>,
    AuthUser(user): AuthUser,
    Path(id): Path<String>,
) -> Result<Json<ProviderDetails>, ApiError> {
    let entry = scoped(&services, &user, &id).await?;
    let disclose = services.vault.configures(&user);
    Ok(Json(services.assembly.details(&entry, disclose).await?))
}

/// The entry's provider for `account`, or for its active account without
/// one.
async fn account_provider(
    services: &Services,
    entry: &ProviderEntry,
    account: Option<&CredentialId>,
) -> Result<Arc<dyn Provider>, ApiError> {
    match account {
        Some(account) => services
            .assembly
            .for_account(entry, account)
            .await?
            .ok_or_else(ApiError::account_not_found),
        None => Ok(services.assembly.provider_for(entry).await?),
    }
}

/// Probes the quota of the account the body names, or of the active one:
/// only a free probe runs, and a family that cannot probe answers its kept
/// snapshot. Cancelling the request stops the probe.
pub(super) async fn quota(
    State(services): State<Arc<Services>>,
    AuthUser(user): AuthUser,
    Path(id): Path<String>,
    OptionalJsonBody(request): OptionalJsonBody<QuotaRequest>,
) -> Result<Json<QuotaAnswer>, ApiError> {
    configures(&services, &user)?;
    let entry = scoped(&services, &user, &id).await?;
    let provider = account_provider(&services, &entry, request.credential_id.as_ref()).await?;
    let Some(quota) = provider.quota() else {
        return Ok(Json(QuotaAnswer { quota: None }));
    };
    let snapshot = match quota.probe().await {
        Ok(snapshot) => Some(snapshot),
        Err(QuotaError::Unsupported) => quota.latest(),
        Err(QuotaError::RequiresInference) => {
            return Err(ApiError::new(
                StatusCode::CONFLICT,
                ErrorCode::QuotaRequiresInference,
                "This provider cannot read its usage without an inference request",
            ));
        }
        Err(error) => {
            return Err(ApiError::new(
                StatusCode::BAD_GATEWAY,
                ErrorCode::QuotaUnavailable,
                error.to_string(),
            ));
        }
    };
    Ok(Json(QuotaAnswer {
        quota: snapshot.map(|snapshot| (*snapshot).clone()),
    }))
}

/// One real request to the model the body names, with the account it names
/// or the active one, on the acting user's shard.
pub(super) async fn test(
    State(services): State<Arc<Services>>,
    State(shards): State<Shards>,
    AuthUser(user): AuthUser,
    Path(id): Path<String>,
    JsonBody(request): JsonBody<TestRequest>,
) -> Result<Json<TestResult>, ApiError> {
    configures(&services, &user)?;
    let entry = scoped(&services, &user, &id).await?;
    let _held = reserve(&services, &entry)?;
    let provider = account_provider(&services, &entry, request.credential_id.as_ref()).await?;
    let model = request.model_id;
    let result = shards
        .of(&user.id)
        .call(move |shard, cancel| async move { shard.test_provider(entry, provider, model, cancel).await })
        .await?;
    Ok(Json(result))
}
